### DISCUSSION TITLE
Starting off a Clean Plate

### DISCUSSION THREAD

**Discussion by hansteg** [ttscoff/bunch-docs#157] [Q&A] [answered]
Hi,

I started testing Bunch yesterday, and warming up to it as it's fast. I'd like to start off a clean plate sometimes, and here is a  things I can't find anything about:

How can I close ALL open apps (except menu bar ones) with one command? My current workaround is to use `!app` and add more to the list as soon as I run across one that doesn't close.

Thanks, Hans

**Comment by ttscoff** [comment, marked as answer]
Hi Hans,

I've never built this into Bunch as it seems impractical to me (or at least not _consistently_ practical), but it's pretty easy to do with AppleScript. I'd entertain the possibility of making it a Bunch feature at some point...

```
tell application "System Events" to set _apps to (name of every process where background only is false)

repeat with _app in _apps
	try
		if _app does not contain "Finder" then
			tell application _app to quit
		end if
	on error
		-- do shell script "Killall " & quoted form of processName
	end try
end repeat
```

You can save that to a file and call it with a `* /path/to/Quit All Apps.applescript`, or use the [Heredoc syntax](https://bunchapp.co/docs/bunch-files/scripts/applescript/#heredoc) to embed the contents directly in a Bunch.

Note that it would quit Finder as well if not for the `if` statement that checks that.  If there are apps more you _don't_ want quit, you'd have to additional logic for them. I can't remember if AppleScript has a delete function that would let you filter a list or not... 

It might fail to quit some apps for various reasons. If there are just a couple apps that don't respond well to it, you can add `!quit` lines for them. 

I left the killall line commented out, but if you wanted to make absolutely sure that everything quit, you could uncomment that line. Use at your own risk, as this will bypass cases where the app failed to quit because you had unsaved changes.

**Comment by hansteg** [comment]
I'll try that, thanks!

On 18 Oct 2021, at 11:48, Brett Terpstra wrote:

> Hi Hans,
>
> I've never built this into Bunch as it seems impractical to me (or at 
> least not _consistently_ practical), but it's pretty easy to do with 
> AppleScript. I'd entertain the possibility of making it a Bunch 
> feature at some point...
>
> ```
> tell application "System Events" to set _apps to (name of every 
> process where background only is false)
>
> repeat with _app in _apps
>     try
>     	if _app does not contain "Finder" then
>     		tell application _app to quit
>     	end if
>     on error
>     	-- do shell script "Killall " & quoted form of processName
>     end try
> end repeat
> ```
>
> You can save that to a file and call it with a `* /path/to/Quit All 
> Apps.applescript`, or use the [Heredoc 
> syntax](https://bunchapp.co/docs/bunch-files/scripts/applescript/#heredoc) 
> to embed the contents directly in a Bunch.
>
> Note that it would quit Finder as well if not for the `if` statement 
> that checks that.  If there are apps more you _don't_ want quit, you'd 
> have to additional logic for them. I can't remember if AppleScript has 
> a delete function that would let you filter a list or not...
>
> It might fail to quit some apps for various reasons. If there are just 
> a couple apps that don't respond well to it, you can add `!quit` lines 
> for them.
>
> I left the killall line commented out, but if you wanted to make 
> absolutely sure that everything quit, you could uncomment that line. 
> Use at your own risk, as this will bypass cases where the app failed 
> to quit because you had unsaved changes.
>
>
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/ttscoff/bunch/discussions/157#discussioncomment-1497024


---------------------
All the best,

Hans

http://tammen.org

...just another worker in rhythms, frequencies and intensities.

**Comment by dwertheimer** [comment]
@ttscoff Thanks to @codedungeon I am a new convert to the app. Thanks for this cool tool.

I pasted the applescript per the instructions above. When I run it, it asks me where submlime_text is. Weird. And somewhat weirder...Sublime text is not in the list of apps! Any idea what's going on?

![Screen Shot 2021-10-22 at 9 31 02 AM](https://user-images.githubusercontent.com/8949588/138491365-0609dcfc-f722-4931-a44e-9ccf8db9d014.jpg)

**Reply by ttscoff** [reply]
Just tested, and yeah, sublime is going to be problematic in this, no scripting dictionary, so it throws Apple Events off. Also had an issue with Forklift's background process which somehow registers as another Dock app. You'll have to make manual exceptions for these with if statements.

```
tell application "System Events" to set _apps to (name of every process where background only is false)

repeat with _app in _apps
	try
		if _app does not contain "Finder" then
			if _app contains "sublime_text" then
				do shell script "killall sublime_text"
			else
				tell application _app to quit
			end if
		end if
	on error
		-- do shell script "Killall " & quoted form of processName
	end try
end repeat
```

This may convince me to add it as a feature to Bunch and just handle this stuff internally. After experimenting with it for a while I've come around to seeing how a clean slate might be nice. Could have a command that took a whitelist of apps _not_ to quit and just kill everything else...

**Comment by dwertheimer** [comment]
Thanks @ttscoff ! I would love to have that “kill all less whitelist” function because I tend to open a ton of apps in the day and then my fans go crazy

**Comment by hansteg** [comment]
Second!

On 22 Oct 2021, at 15:34, David Wertheimer wrote:

> Thanks @ttscoff ! I would love to have that “kill all less 
> whitelist” function because I tend to open a ton of apps in the day 
> and then my fans go crazy
>
> -- 
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/ttscoff/bunch/discussions/157#discussioncomment-1523089


---------------------
All the best,

Hans

http://tammen.org

...just another worker in rhythms, frequencies and intensities.

**Comment by beginnerninja** [comment]
Thirded! 👍

**Comment by ttscoff** [comment]
This has been added in Bunch Beta (132). Update to the latest and check the docs:

https://bunchapp.co/docs/bunch-files/commands/quit-everything/

**Comment by beginnerninja** [comment]
@ttscoff - it works amazingly well! Thank you!

**Comment by dwertheimer** [comment]
Awesome!

**Comment by hansteg** [comment]
excellent, works like a charm!

On 28 Oct 2021, at 18:24, beginnerninja wrote:

> @ttscoff - it works amazingly well! Thank you!

Hans
